ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -58,12 +78,78 @@ func NewRayServiceReconciler(mgr manager.Manager) *RayServiceReconciler {
-	_ = logf.FromContext(ctx)
+func (r *RayServiceReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("rayservice", request.NamespacedName)
+	log.Info("reconciling RayService", "service NamespacedName", request.NamespacedName)
+
+	// Get serving cluster instance
+	var err error
+	var rayServiceInstance *rayv1alpha1.RayService
+
+	if rayServiceInstance, err = r.getRayServiceInstance(request); err != nil {
+		err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_TO_GET_RAYSERVICE, err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var rayClusterInstance *rayv1alpha1.RayCluster
+	if rayClusterInstance, err = r.getOrCreateRayClusterInstance(rayServiceInstance); err != nil {
+		err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_TO_GET_OR_CREATE_RAYCLUSTER, err)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	rayServiceInstance.Status.RayClusterStatus = rayClusterInstance.Status
+
+	var clientURL string
+	if clientURL, err = r.fetchDashboardURL(rayClusterInstance); err != nil || clientURL == "" {
+		err = r.updateState(rayServiceInstance, rayv1alpha1.WAIT_FOR_DASHBOARD, err)
+		return ctrl.Result{}, err
+	}
+
+	rayDashboardClient := utils.RayDashboardClient{}
+	rayDashboardClient.InitClient(clientURL)
+
+	shouldUpdate := r.checkIfNeedSubmitServeDeployment(rayServiceInstance, request)
+
+	if shouldUpdate {
+		if err = r.updateServeDeployment(rayServiceInstance, rayDashboardClient, request); err != nil {
+			err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_SERVE_DEPLOY, err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	var isHealthy bool
+	if isHealthy, err = r.getAndCheckServeStatus(rayServiceInstance, rayDashboardClient); err != nil {
+		err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_GET_SERVE_DEPLOYMENT_STATUS, err)
+		return ctrl.Result{}, err
+	}
+
+	log.Info("Check serve health", "isHealthy", isHealthy)
+
+	if isHealthy {
+		rayServiceInstance.Status.ServiceStatus = rayv1alpha1.RUNNING
+		if err := r.Status().Update(context.Background(), rayServiceInstance); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		rayServiceInstance.Status.ServiceStatus = rayv1alpha1.RESTARTING
+		if err := r.Status().Update(context.Background(), rayServiceInstance); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		delete(r.ServeDeploymentConfigMap, request.NamespacedName)
 
-	// TODO(user): your logic here
+		// restart raycluster
+		if err := r.Delete(ctx, rayClusterInstance); err != nil {
+			err = r.updateState(rayServiceInstance, rayv1alpha1.FAIL_DELETE_RAYCLUSTER, err)
+			return ctrl.Result{}, err
+		}
 
-	return ctrl.Result{}, nil
+		log.V(1).Info("Deleted rayCluster for rayService run", "rayCluster", rayClusterInstance)
+		// Wait a while for the cluster delete
+		return ctrl.Result{RequeueAfter: RayServiceRestartRequeueDuration}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: RayServiceDefaultRequeueDuration}, nil
```

Comment:
@simon-mo re: a question from above, I think `RayServiceDefaultRequeueDuration` would be the backoff after a failed reconcile iteration

Reply:
If we set it, then we use the RayServiceDefaultRequeueDuration value. Otherwise, return error not null will trigger an exponential backoff manner   wait.